File: metisfl/controller/core/learner_manager.cc
#include "metisfl/controller/core/learner_manager.h"

namespace metisfl::controller {
using google::protobuf::util::TimeUtil;

// Constructor
LearnerManager::LearnerManager()
    : learners_(),
      train_params_(),
      eval_params_(),
      learners_mutex_(),
      scheduling_pool_(2),
      train_tasks_cq_(),
      eval_tasks_cq_() {
  std::thread run_tasks_digest_t_(&LearnerManager::DigestTrainResponses, this);
  std::thread eval_tasks_digest_t_(&LearnerManager::DigestEvaluateResponses,
                                   this);
  run_tasks_digest_t_.detach();
  eval_tasks_digest_t_.detach();
}

// Public methods
absl::StatusOr<std::string> LearnerManager::AddLearner(
    const Learner &learner, bool is_semi_sync,
    const std::string &scaling_factor) {
  std::lock_guard<std::mutex> learners_guard(learners_mutex_);

  std::string learner_id =
      GenerateLearnerId(learner.hostname(), learner.port());

  if (learners_.contains(learner_id)) {
    return absl::AlreadyExistsError("Learner has already joined.");
  }
  learners_[learner_id] = learner;

  TrainParams train_params;
  if (is_semi_sync) {
    train_params.add_metadata("processing_ms_per_batch");
    train_params.add_metadata("processing_ms_per_epoch");
  }
  if (scaling_factor == "NumTrainingExamples")
    train_params.add_metadata("num_training_examples");
  if (scaling_factor == "NumCompletedBatches")
    train_params.add_metadata("num_completed_batches");

  train_params_[learner_id] = train_params;

  return learner_id;
}

std::vector<std::string> LearnerManager::GetLearnerIds() const {
  std::vector<std::string> learner_ids;
  for (const auto &[key, learner] : learners_) {
    learner_ids.push_back(key);
  }
  return learner_ids;
}

absl::Status LearnerManager::RemoveLearner(const std::string &learner_id) {
  std::lock_guard<std::mutex> learners_guard(learners_mutex_);

  if (!learners_.contains(learner_id)) {
    return absl::NotFoundError("Learner does not exist.");
  }

  learners_.erase(learner_id);
  train_params_.erase(learner_id);
  eval_params_.erase(learner_id);

  return absl::OkStatus();
}

void LearnerManager::ScheduleTrain(const std::vector<std::string> &learner_ids,
                                   const Model &model) {
  std::lock_guard<std::mutex> learners_guard(learners_mutex_);
  for (const auto &learner_id : learner_ids) {
    scheduling_pool_.push_task(
        [this, learner_id, model] { SendTrainAsync(learner_id, model); });
  }
}

void LearnerManager::ScheduleEvaluate(
    const std::vector<std::string> &learner_ids, const Model &model) {
  std::lock_guard<std::mutex> learners_guard(learners_mutex_);
  for (const auto &learner_id : learner_ids) {
    scheduling_pool_.push_task(
        [this, learner_id, model] { SendEvaluateAsync(learner_id, model); });
  }
}

void LearnerManager::Shutdown() {
  train_tasks_cq_.Shutdown();
  eval_tasks_cq_.Shutdown();
  scheduling_pool_.wait_for_tasks();
}

void LearnerManager::UpdateTrainResults(const Task &task,
                                        const std::string &learner_id,
                                        const TrainResults &results) {
  std::lock_guard<std::mutex> learners_guard(learners_mutex_);
  auto task_id = task.id();
  train_results_[task_id] = results;
  latest_train_results_[learner_id] = results;
  *tasks_[task_id].mutable_received_at() = task.received_at();
  *tasks_[task_id].mutable_completed_at() = task.completed_at();
}

absl::flat_hash_map<std::string, int> LearnerManager::GetNumTrainingExamples(
    const std::vector<std::string> &learner_ids) {
  absl::flat_hash_map<std::string, int> num_training_examples;

  for (const auto &learner_id : learner_ids) {
    num_training_examples[learner_id] = (int)latest_train_results_[learner_id]
                                            .metadata()
                                            .find("num_training_examples")
                                            ->second;
  }
  return num_training_examples;
}

absl::flat_hash_map<std::string, int> LearnerManager::GetNumCompletedBatches(
    const std::vector<std::string> &learner_ids) {
  absl::flat_hash_map<std::string, int> num_completed_batches;

  for (const auto &learner_id : learner_ids) {
    num_completed_batches[learner_id] = (int)latest_train_results_[learner_id]
                                            .metadata()
                                            .find("num_completed_batches")
                                            ->second;
  }

  return num_completed_batches;
}

void LearnerManager::UpdateTrainParams(
    const std::vector<std::string> &learner_ids, const int semi_sync_lambda) {
  // Finds the slowest learner.
  float ms_per_epoch_slowest = std::numeric_limits<float>::min();
  for (const auto &learner_id : learner_ids) {
    auto processing_ms_per_epoch = train_results_[learner_id]
                                       .metadata()
                                       .find("processing_ms_per_epoch")
                                       ->second;
    if (processing_ms_per_epoch > ms_per_epoch_slowest) {
      ms_per_epoch_slowest = processing_ms_per_epoch;
    }
  }

  // Calculates the allowed time for training.
  float t_max = static_cast<float>(semi_sync_lambda) * ms_per_epoch_slowest;

  // Updates the task templates based on the slowest learner.
  for (const auto &learner_id : learner_ids) {
    auto processing_ms_per_batch = train_results_[learner_id]
                                       .metadata()
                                       .find("processing_ms_per_batch")
                                       ->second;
    if (processing_ms_per_batch == 0) {
      // FIXME: Better error handling.
      LOG(ERROR) << "Processing ms per batch is zero. Setting to 1.";
      processing_ms_per_batch = 1;
    }
    int num_local_updates = std::ceil(t_max / processing_ms_per_batch);

    train_params_[learner_id].set_num_local_updates(num_local_updates);
  }
}

// Private methods
LearnerStub LearnerManager::CreateLearnerStub(const std::string &learner_id) {
  auto hostname = learners_[learner_id].hostname();
  auto port = learners_[learner_id].port();
  auto target = absl::StrCat(hostname, ":", port);
  auto &root_certificate = learners_[learner_id].root_certificate_bytes();

  auto ssl_creds = grpc::InsecureChannelCredentials();

  if (!root_certificate.empty()) {
    grpc::SslCredentialsOptions ssl_opts;
    ssl_opts.pem_root_certs = root_certificate;
    ssl_creds = grpc::SslCredentials(ssl_opts);
  }

  auto channel = grpc::CreateChannel(target, ssl_creds);
  return LearnerService::NewStub(channel);
}

bool LearnerManager::ValidateLearner(const std::string &learner_id) const {
  return learners_.contains(learner_id);
}

void LearnerManager::SendTrainAsync(const std::string &learner_id,
                                    const Model &model) {
  auto task_id = GenerateRadnomId();
  tasks_[task_id] = Task();
  *tasks_[task_id].mutable_id() = task_id;
  *tasks_[task_id].mutable_learner_id() = learner_id;
  *tasks_[task_id].mutable_sent_at() = TimeUtil::GetCurrentTime();

  TrainRequest request;
  *request.mutable_task() = tasks_[task_id];
  *request.mutable_model() = model;
  *request.mutable_params() = train_params_[learner_id];

  auto *call = new AsyncLearnerRunTaskCall;
  auto &cq = train_tasks_cq_;
  auto learner_stub = CreateLearnerStub(learner_id);

  call->learner_id = learner_id;
  call->response_reader =
      learner_stub->PrepareAsyncTrain(&call->context, request, &cq);
  call->response_reader->StartCall();
  call->response_reader->Finish(&call->reply, &call->status, (void *)call);
}

void LearnerManager::DigestTrainResponses() {
  void *got_tag;
  bool ok = false;
  auto &cq_ = train_tasks_cq_;
  while (cq_.Next(&got_tag, &ok)) {
    auto *call = static_cast<AsyncLearnerRunTaskCall *>(got_tag);
    GPR_ASSERT(ok);

    if (call) {
      if (!call->status.ok()) {
        LOG(ERROR) << "Train RPC request to learner: " << call->learner_id
                   << " failed with error: " << call->status.error_message();
      }
    }
    delete call;
  }
}

void LearnerManager::SendEvaluateAsync(const std::string &learner_id,
                                       const Model &model) {
  auto task_id = GenerateRadnomId();
  tasks_[task_id] = Task();
  *tasks_[task_id].mutable_id() = task_id;
  *tasks_[task_id].mutable_learner_id() = learner_id;
  *tasks_[task_id].mutable_sent_at() = TimeUtil::GetCurrentTime();

  EvaluateRequest request;
  *request.mutable_task() = tasks_[task_id];
  *request.mutable_model() = model;
  *request.mutable_params() = eval_params_[learner_id];

  auto *call = new AsyncLearnerEvalCall;
  auto &cq = eval_tasks_cq_;
  auto learner_stub = CreateLearnerStub(learner_id);

  call->learner_id = learner_id;
  call->response_reader =
      learner_stub->PrepareAsyncEvaluate(&call->context, request, &cq);
  call->response_reader->StartCall();
  call->response_reader->Finish(&call->reply, &call->status, (void *)call);
}

void LearnerManager::DigestEvaluateResponses() {
  void *got_tag;
  bool ok = false;
  auto &cq_ = eval_tasks_cq_;
  while (cq_.Next(&got_tag, &ok)) {
    auto *call = static_cast<AsyncLearnerEvalCall *>(got_tag);
    GPR_ASSERT(ok);

    if (call) {
      if (call->status.ok()) {
        const std::string &task_id = call->reply.task().id();
        evaluation_results_[task_id] = call->reply.results();
        *tasks_[task_id].mutable_received_at() =
            call->reply.task().received_at();
        *tasks_[task_id].mutable_completed_at() =
            call->reply.task().completed_at();
      } else {
        LOG(ERROR) << "EvaluateModel RPC request to learner: "
                   << call->learner_id
                   << " failed with error: " << call->status.error_message();
      }
    }
    delete call;
  }
}

}  // namespace metisfl::controller

File: metisfl/server/core/learner_manager.py


from typing import Dict, List, Optional
from metisfl.common.client import get_client
from metisfl.common.types import ClientParams

from metisfl.proto import (controller_pb2, controller_pb2_grpc, learner_pb2, learner_pb2_grpc,
                           model_pb2, service_common_pb2)


def get_learner_id(
    hostname: str,
    port: int,
) -> str:
    """Gets the learner id from the hostname and port."""
    return f"{hostname}:{port}"


class LearnerManager:

    # learner_id -> {}
    learners: Dict[str, controller_pb2.Learner] = {}
    client_params: Dict[str, ClientParams] = {}
    train_params: learner_pb2.TrainParams = None
    eval_params: learner_pb2.EvalParams = None

    # task_id -> {}
    tasks: Dict[str, learner_pb2.Task] = {}
    train_results: Dict[str, controller_pb2.TrainResult] = {}
    eval_results: Dict[str, learner_pb2.EvalResult] = {}

    # learner_id -> {}
    last_train_results: Dict[str, controller_pb2.TrainResult] = {}
    num_training_examples: Dict[str, int] = {}

    def __init__(self, learner):
        self.learner = learner

    def add_learner(self, learner: controller_pb2.Learner) -> str:
        """Adds a learner to the controller.

        Parameters
        ----------
        learner : controller_pb2.Learner
            The learner to be added.
        """
        learner_id = get_learner_id(
            hostname=learner.hostname,
            port=learner.port,
        )

        if learner_id in self.learners:
            raise ValueError(f"Learner {learner_id} already exists.")

        self.learners[learner_id] = learner
        self.client_params[learner_id] = ClientParams(
            hostname=learner.hostname,
            port=learner.port,
            root_certificate=learner.root_certificate_bytes,
        )

        return learner_id

    def remove_learner(self, learner_id: str) -> None:
        """Removes a learner from the controller.

        Parameters
        ----------
        learner_id : str
            The learner id.
        """
        if learner_id not in self.learners:
            raise ValueError(f"Learner {learner_id} does not exist.")

        self.learners.pop(learner_id)
        self.clients.pop(learner_id)

    def get_client(self, learner_id: str):
        return get_client(
            stub_class=controller_pb2_grpc.ControllerServiceStub,
            client_params=self.client_params[learner_id],
        )

    def schedule_train(
        self,
        task: learner_pb2.Task,
        model: model_pb2.Model,
        train_params: learner_pb2.TrainParams,
        request_retries: Optional[int] = 1,
        request_timeout: Optional[int] = None,
        block: Optional[bool] = False
    ) -> service_common_pb2.Ack:

        with self.get_client() as client:

            stub: learner_pb2_grpc.LearnerServiceStub = client[0]
            schedule = client[1]

            def _request(_timeout=None):

                request = controller_pb2.TrainRequest(
                    task=task,
                    model=model,
                    params=train_params
                )

                return stub.Train(request, timeout=_timeout)

            return schedule(_request, request_retries, request_timeout, block)

    def schedule_evaluate(
        self,
        task: learner_pb2.Task,
        model: model_pb2.Model,
        eval_params: learner_pb2.EvaluationParams,
        request_retries: Optional[int] = 1,
        request_timeout: Optional[int] = None,
        block: Optional[bool] = False
    ) -> service_common_pb2.Ack:

        with self.get_client() as client:

            stub: learner_pb2_grpc.LearnerServiceStub = client[0]
            schedule = client[1]

            def _request(_timeout=None):

                request = controller_pb2.EvaluateRequest(
                    task=task,
                    model=model,
                    params=eval_params
                )

                return stub.Evaluate(request, timeout=_timeout)

            return schedule(_request, request_retries, request_timeout, block)

    def shutdown_client(self):
        """Shuts down the client."""
        with self.get_client() as client:
            client[2].shutdown()

    def get_num_training_examples(self, learner_ids: List[str]) -> Dict[str, int]:
        """Gets the number of training examples for the learners.

        Parameters
        ----------
        learner_ids : List[str]
            The learner ids.

        Returns
        -------
        Dict[str, int]
            The number of training examples for each learner.
        """
        num_training_examples = {}
        for learner_id in learner_ids:
            num_training_examples[learner_id] = self.num_training_examples.get(
                learner_id, 0
            )

    def get_num_completed_batches(self, learner_ids: List[str]) -> Dict[str, int]:
        """Gets the number of completed batches for the learners.

        Parameters
        ----------
        learner_ids : List[str]
            The learner ids.

        Returns
        -------
        Dict[str, int]
            The number of completed batches for each learner.
        """
        num_completed_batches = {}
        for learner_id in learner_ids:
            num_completed_batches[learner_id] = self.last_train_results.get(
                learner_id, 0
            )  # FIXME: this is not correct

    def get_learner_ids(self) -> List[str]:
        """Gets the learner ids.

        Returns
        -------
        List[str]
            The learner ids.
        """
        return list(self.learners.keys())

    def get_learner_id(self, task_id: str) -> str:
        """Gets the learner id of a task."""
        return self.tasks[task_id].learner_id

    def update_train_result(
        self,
        task: learner_pb2.Task,
        learner_id: str,
        train_results: controller_pb2.TrainResults,
    ) -> None:
        """Updates the train result of a learner.

        Parameters
        ----------
        task : learner_pb2.Task
            The newly completed task.
        learner_id : str
            The learner id.
        train_result : controller_pb2.TrainResults
            The train result of the task.
        """
        task_id = task.id
        self.train_results[task_id] = train_results
        self.last_train_results[learner_id] = train_results
        self.tasks[task_id].received_at = task.received_at
        self.tasks[task_id].completed_at = task.completed_at
